### DISCUSSION TITLE
Expunging an object loaded from a relationship is memoized and consecutive loads return same expunged objects.

### DISCUSSION THREAD

**Discussion by yoshpe** [sqlalchemy/sqlalchemy#10110] [Usage Questions]
Hey, following this 
https://stackoverflow.com/questions/76641620/sqlalchemy-parent-instance-not-bound-to-a-session-even-though-it-should-be

I discovered in my code, a flow that loads models from a relationship, and then expunges them from the session.
Afterwards fetching from that relationship again returned the same expunged objects.

Simplified code
```python
def func1():
    # This was hidden deeply within the call to func1 (it was actually further down the stack)
    m = session.query(Model).get(...)
    for other in m.other_relationship:
        # Let's expunge these objects because we need to modify them, serialize them, and whatever, outside of the session
        session.expunge(other)

def func2():
    m = session.query(Model).get(...)
    for other in m.other_relationship:
        # Not obvious why was other it detached from session??
        assert other not in session

def func3():
    func1()
    # Calling func2 from other flows in the code works as intended, but it fails from func3's flow because of func1's behavior
    func2()
```

Fetching all of `m.other_relationship` explicitly using their primary keys return new objects that are in the session,
however iterating through `m.other_relationship` will continue to return the expunged objects.

When thinking of sqlalchemy's potential memoization, this may not be surprising.
Fetching Model the second time (in `func2`) returns the same identical object
i.e.
```python
m1 = session.query(Model).get(...)
m2 = session.query(Model).get(...)
m1 is m2
```

And then mutating the `other_relationship` objects is done to their reference which is probably memoized<br/>
However I couldn't find this memoization while I was debugging and digging internally into InstrumentedList and InstrumentedAttribute (tried to breakpoint in many places but it didn't hit, also I wasn't able to "step into" the relationship's iterator)<br/>
The debugging of this was super tricky, as I wasn't able to easily figure out where and why the relationship was already cached.

Is this the intended behavior? 

Ultimately I fixed the code by using `copy.deepcopy` on the other_relationship's objects in `func1`

**Comment by zzzeek** [comment]
hi -

first off, copy.deepcopy is not recommended for SQLAlchemy objects as there is an `_sa_instance_state` member referring to `InstanceState` and that is not designed to be copied with deepcopy.   There are library supported ways of building transient and detached objects: https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.make_transient https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.make_transient_to_detached 

second....

> Is this the intended behavior?

that...expunging the objects in a collection would expunge the parent object?   Yes if you have the [default cascade rules set up](https://docs.sqlalchemy.org/en/20/orm/cascades.html) and these objects in the collection have a [backref or back_populates to that parent](https://docs.sqlalchemy.org/en/20/tutorial/orm_related_objects.html#tutorial-orm-related-objects).

you can turn off the "expunge" cascade for that back-relationship, however I would also not recommend getting into changing those things, as using `expuge()` already is an awkward pattern, and especially using it with the members of a collection.  that is  awkward and I dont see what use case is accomplished by that.

> Let's expunge these objects because we need to modify them, 

modifying you would normally do with the objects attached, so that the changes you make can be persisted back to the database in the transaction you have.

> serialize them, 

serialization means you are generating a string representation of those objects that can be sent over a network connection.   There's no need to detach the objects in this case, they will produce the same serialization whether or not they are attached to a session.

> and whatever, outside of the session

so...make a new `Session` that's specific to this non-session operation, emit a SELECT directly for those objects, then close that `Session`.   it's awkward that you seem to have an operation that is doing two separate things, one is loading a parent object for some purpose which you want to remain attached, and the other is loading some objects that are directly associated with that parent object and those you want to remain detached.   

as mentioned above, you can use [make_transient](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.make_transient) as kind of a large, blunt hammer that will change the `InstanceState` flags on those objects so they will appear as though you just constructed them.  you will want to `session.expire()` the parent object's collection that you are pulling them from, as they otherwise would still be in there and would definitely cause problems with that parent going forward.
